src/cloud-api-adaptor/test/provisioner/gcp/cluster.go

```diff
@@ -0,0 +1,346 @@
+	"context"
+	"fmt"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strconv"
+	"time"
+
+	log "github.com/sirupsen/logrus"
+	"google.golang.org/api/container/v1"
+	"google.golang.org/api/googleapi"
+	"google.golang.org/api/option"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/clientcmd"
+	"k8s.io/client-go/util/retry"
+	kconf "sigs.k8s.io/e2e-framework/klient/conf"
+
+	appsv1 "k8s.io/api/apps/v1"
+	"k8s.io/apimachinery/pkg/util/yaml"
+)
+
+// GKECluster implements the basic GKE Cluster client operations.
+type GKECluster struct {
+	clusterName        string
+	clusterVersion     string
+	clusterMachineType string
+	credentials        string
+	nodeCount          int64
+	ProjectID          string
+	Zone               string
+	cluster            *container.Cluster
+}
+
+// NewGKECluster creates a new GKECluster with the given properties
+func NewGKECluster(properties map[string]string) (*GKECluster, error) {
+	defaults := map[string]string{
+		"cluster_name":         "e2e-peer-pods",
+		"cluster_version":      "1.31.4-gke.1256000",
+		"cluster_machine_type": "n1-standard-1",
+		"node_count":           "2",
+	}
+
+	for key, value := range properties {
+		defaults[key] = value
+	}
+
+	requiredFields := []string{"project_id", "credentials", "zone"}
+	for _, field := range requiredFields {
+		if _, ok := defaults[field]; !ok {
+			return nil, fmt.Errorf("%s is required", field)
+		}
+	}
+
+	nodeCount, err := strconv.ParseInt(defaults["node_count"], 10, 64)
+	if err != nil {
+		return nil, fmt.Errorf("invalid node_count: %v", err)
+	}
+
+	return &GKECluster{
+		clusterName:        defaults["cluster_name"],
+		clusterVersion:     defaults["cluster_version"],
+		clusterMachineType: defaults["cluster_machine_type"],
+		credentials:        defaults["credentials"],
+		nodeCount:          nodeCount,
+		ProjectID:          defaults["project_id"],
+		Zone:               defaults["zone"],
+		cluster:            nil,
+	}, nil
+}
+
+// Apply basic labels to worker nodes
+func (g *GKECluster) ApplyNodeLabels(ctx context.Context) error {
+	kubeconfigPath, err := g.GetKubeconfigFile(ctx)
+	if err != nil {
+		return err
+	}
+
+	config, err := clientcmd.BuildConfigFromFlags("", kubeconfigPath)
+	if err != nil {
+		return fmt.Errorf("failed to build kubeconfig: %v", err)
+	}
+
+	clientset, err := kubernetes.NewForConfig(config)
+	if err != nil {
+		return fmt.Errorf("failed to create clientset: %v", err)
+	}
+
+	nodes, err := clientset.CoreV1().Nodes().List(ctx, metav1.ListOptions{})
+	if err != nil {
+		return fmt.Errorf("failed to list nodes: %v", err)
+	}
+
+	for _, node := range nodes.Items {
+		err := retry.RetryOnConflict(retry.DefaultRetry, func() error {
+			n, err := clientset.CoreV1().Nodes().Get(ctx, node.Name, metav1.GetOptions{})
+			if err != nil {
+				return fmt.Errorf("failed to get node: %v", err)
+			}
+
+			n.Labels["node.kubernetes.io/worker"] = ""
+			n.Labels["node-role.kubernetes.io/worker"] = ""
```

Comment:
`node-role.kubernetes.io/worker` is not needed.

Also, how do you know that you are not applying the label to the controllers nodes?

Reply:
Since is GKE, the control plane is managed by Google and doesn’t show up in clientset.CoreV1().Nodes().List(). During my tests, it wasn’t there.

About the label, fixing with a force push.

Reply:
ok, got it. thanks!